kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/JsonUtil.java

```diff
@@ -79,4 +117,449 @@ public static JsonNode parse(InputStream in) {
+        case ARRAY:
+          if (datum.isArray()) {
+            return schema;
+          }
+          break;
+        case BOOLEAN:
+          if (datum.isBoolean()) {
+            return schema;
+          }
+          break;
+        case FLOAT:
+          if (datum.isFloat() || datum.isInt()) {
+            return schema;
+          }
+          break;
+        case DOUBLE:
+          if (datum.isDouble() || datum.isFloat() ||
+              datum.isLong() || datum.isInt()) {
+            return schema;
+          }
+          break;
+        case INT:
+          if (datum.isInt()) {
+            return schema;
+          }
+          break;
+        case LONG:
+          if (datum.isLong() || datum.isInt()) {
+            return schema;
+          }
+          break;
+        case STRING:
+          if (datum.isTextual()) {
+            return schema;
+          }
+          break;
+        case ENUM:
+          if (datum.isTextual() && schema.hasEnumSymbol(datum.textValue())) {
+            return schema;
+          }
+          break;
+        case BYTES:
+        case FIXED:
+          if (datum.isBinary()) {
+            return schema;
+          }
+          break;
+        case NULL:
+          if (datum == null || datum.isNull()) {
+            return schema;
+          }
+          break;
+        default: // UNION or unknown
+          throw new IllegalArgumentException("Unsupported schema: " + schema);
+      }
+    }
+    throw new DatasetException(String.format(
+        "Cannot resolve union: %s not in %s", datum, schemas));
+  }
+
+  public static Schema inferSchema(InputStream incoming, final String name,
+                                   int numRecords) {
+    Iterator<Schema> schemas = Iterators.transform(parser(incoming),
+        new Function<JsonNode, Schema>() {
+          @Override
+          public Schema apply(JsonNode node) {
+            return inferSchema(node, name);
+          }
+        });
+
+    if (!schemas.hasNext()) {
+      return null;
+    }
+
+    Schema result = schemas.next();
+    for (int i = 1; schemas.hasNext() && i < numRecords; i += 1) {
+      result = SchemaUtil.merge(result, schemas.next());
+    }
+
+    return result;
+  }
+
+  public static Schema inferSchema(JsonNode node, String name) {
+    return visit(node, new JsonSchemaVisitor(name));
+  }
+
+  public static Schema inferSchemaWithMaps(JsonNode node, String name) {
+    return visit(node, new JsonSchemaVisitor(name).useMaps());
+  }
+
+  private static class JsonSchemaVisitor extends JsonTreeVisitor<Schema> {
+
+    private static final Joiner DOT = Joiner.on('.');
+    private final String name;
+    private boolean objectsToRecords = true;
+
+    public JsonSchemaVisitor(String name) {
+      this.name = name;
+    }
+
+    public JsonSchemaVisitor useMaps() {
+      this.objectsToRecords = false;
+      return this;
+    }
+
+    @Override
+    public Schema object(ObjectNode _, Map<String, Schema> fields) {
```

Comment:
We shouldn't use `_` as an identifier as Java may make it a keyword in a future release.

From http://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.27.1:

**It is a compile-time error if a lambda parameter has the name `_` (that is, a single underscore character).**

*The use of the variable name `_` in any context is discouraged. Future versions of the Java programming language may reserve this name as a keyword and/or give it special semantics.*